lib/msal-browser/src/interaction_client/RedirectClient.ts

```diff
@@ -331,10 +344,55 @@ export class RedirectClient extends StandardInteractionClient {
                 (e as AuthError).setCorrelationId(this.correlationId);
                 serverTelemetryManager.cacheFailedRequest(e);
             }
+
+            if (
+                e instanceof ServerError &&
+                e.errorCode === BrowserConstants.INVALID_GRANT_ERROR
+            ) {
+                this.performanceClient.addFields(
+                    {
+                        retryError: e.errorCode,
+                    },
+                    this.correlationId
+                );
+
+                const requestRetried = this.browserStorage.getRequestRetried(
+                    this.correlationId
+                );
+
+                if (requestRetried) {
+                    this.logger.error(
+                        "Retried request already detected. Throwing error."
+                    );
+                    this.browserStorage.removeRequestRetried(
+                        this.correlationId
+                    );
+                    throw e;
+                }
+
+                const redirectRequest =
+                    this.browserStorage.getCachedRedirectRequest();
+                if (!redirectRequest) {
+                    this.logger.error(
+                        `Unable to retry. Please retry with redirect request and correlationId: ${this.correlationId}`
+                    );
+                    this.browserStorage.setRequestRetried(this.correlationId);
```

Comment:
If the application doesn't reuse the same correlationId will this still get cleaned up somewhere or will it persist forever?

We'll want to make sure these new cache entries are cleaned up no matter what after handleRedirectPromise is finished

Reply:
Good point. I changed the removal logic and added it in for both success and failure cases.